presto-hive/src/main/java/io/prestosql/plugin/hive/HiveApplyProjectionUtil.java

```diff
@@ -0,0 +1,225 @@
+                }
+                else {
+                    return false;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    static ProjectedColumnRepresentation createProjectedColumnRepresentation(ConnectorExpression expression)
+    {
+        List<ConnectorExpression> orderedPrefixes = createOrderedPrefixes(expression);
+        checkArgument(orderedPrefixes.size() > 0, "invalid ordered prefixes");
+
+        Variable variable = (Variable) orderedPrefixes.get(0);
+        ImmutableList.Builder<Integer> dereferenceIntegers = ImmutableList.builder();
+
+        for (int i = 1; i < orderedPrefixes.size(); i++) {
+            FieldDereference dereference = (FieldDereference) orderedPrefixes.get(i);
+            dereferenceIntegers.add(dereference.getField());
+        }
+
+        return new ProjectedColumnRepresentation(variable, dereferenceIntegers.build());
+    }
+
+    /**
+     * Return a list of all prefixes for a variable OR a dereference chain ordered in the increasing order of their
+     * lengths. For example, this method would return a list of {@link ConnectorExpression} representing ["a", "a.b", "a.b.c"]
+     * for a {@link ConnectorExpression} "a.b.c".
+     */
+    static List<ConnectorExpression> createOrderedPrefixes(ConnectorExpression expression)
+    {
+        ImmutableList.Builder<ConnectorExpression> prefixes = ImmutableList.builder();
+
+        ConnectorExpression current = expression;
+
+        while (true) {
+            checkArgument(expression instanceof FieldDereference || expression instanceof Variable);
+            prefixes.add(current);
+
+            if (current instanceof Variable) {
+                break;
+            }
+            else {
+                current = ((FieldDereference) current).getTarget();
+            }
+        }
+
+        return prefixes.build().reverse();
+    }
+
+    /**
+     * Replace all connector expressions with variables as given by {@param expressionToVariableMappings} in a top down manner.
+     * i.e. if the replacement occurs for the parent, the children will not be visited.
+     */
+    public static ConnectorExpression replaceWithNewVariables(ConnectorExpression expression, Map<ConnectorExpression, Variable> expressionToVariableMappings)
+    {
+        if (expressionToVariableMappings.containsKey(expression)) {
+            return expressionToVariableMappings.get(expression);
+        }
+
+        if (expression instanceof FieldDereference) {
+            ConnectorExpression newTarget = replaceWithNewVariables(((FieldDereference) expression).getTarget(), expressionToVariableMappings);
+            return new FieldDereference(expression.getType(), newTarget, ((FieldDereference) expression).getField());
+        }
+
+        if (expression instanceof Variable) {
+            return expression;
+        }
+
+        if (expression instanceof Constant) {
+            return expression;
+        }
```

Comment:
I had intentionally kept it for readability, like switch/case statements so that it's easier to add logic for future additions to `ConnectorExpression` family. But it's not critical, removed it in the update.